### PR TITLE
Introduce previously_new_record? on ActiveRecord::Base

### PR DESCRIPTION
**Pull request by tomafro**
### Summary

Adds a `previously_new_record?` method to Active Record models, returning true if prior to the last save the model was a new record. The name `previously_new_record?` was chosen to mirror methods like `x_previously_changed?`.

Take a table with a `status` column with default value `published`:
```ruby
create_table :examples do |table|
  table.string :status, default: 'published'
end
```
If we want to indicate whether it's the first time the record has been saved with the status `published`, we can't use `status_previously_changed?`, as it rightly doesn't register a change from the default. We can check `id_previously_changed?` instead, but this both isn't as descriptive, and can technically be triggered in other situations. `record.update id: 999, status: 'published'` is legal if unwise code.

A specific `status_previously_changed?` method is both clearer and more correct.
```ruby
class Example < ActiveRecord::Base
  after_save_commit :broadcast_publication, if: :previously_published?

  # Implementation without new previously_new_record? method
  def previously_published?
    status == "published" && (status_previously_changed? || id_previously_changed?)
  end

  # Implementation with
  def previously_published?
    status == "published" && (status_previously_changed? || previously_new_record?)
  end
end
```

If this addition passes muster, @dhh has also suggested adding similar methods to enums, so if `status` was an enum rails would provide `previously_published?` itself. I would add this in a separate PR.

### Other Information

We're using this right now at Basecamp, and @dhh asked me to bring it over to rails.

I did consider an implementation that used an alias (or similar) to `id_previously_changed?` but didn't for the following minor reasons:

- We can't simply alias `id_previously_changed?`, we'd need to accommodate cases where the model uses a different primary key column so it's a tiny bit more complicated.

- It wouldn't work for composite primary keys, which all though rails doesn't natively support, I know some people use.

- It's possible (though obviously unwise) to change the primary key of an existing model, using something like `record.update id: 999, status: 'published'`. Given how many people use rails, I'm sure this will bite somebody.

For the implementation I've tried to mirror existing similar code and tests as much as possible, but I'm not deeply familiar with the code, so very happy to change.

### CODE REVIEW

**Review by eileencodes — APPROVED** [review]
This looks good to me. Can you add a changelog entry?

File: activerecord/lib/active_record/persistence.rb [hunk 1]
```diff
@@ -428,6 +428,14 @@ def new_record?
       @new_record
     end
 
+    # Returns true if this object was just created -- that is, prior to the last
+    # save, the object didn't exist in the database and new_record? would have
+    # returned true.
+    def previously_new_record?
+      sync_with_transaction_state
```
Comment (kamipo) [inline activerecord/lib/active_record/persistence.rb]: Can you sync the line with one above and below methods?

```suggestion
      sync_with_transaction_state if @transaction_state&.finalized?
```

https://github.com/rails/rails/blob/251037efb383c7cfeba98cdc7e6fc8a7f12933a9/activerecord/lib/active_record/persistence.rb#L426-L427
https://github.com/rails/rails/blob/251037efb383c7cfeba98cdc7e6fc8a7f12933a9/activerecord/lib/active_record/persistence.rb#L440-L441
https://github.com/rails/rails/blob/251037efb383c7cfeba98cdc7e6fc8a7f12933a9/activerecord/lib/active_record/persistence.rb#L447-L448

**Comment by tomafro** [comment]
Thanks for the feedback @kamipo and @eileencodes. I've made the changes you both suggested.

File: activerecord/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
+*   Add `ActiveRecord::Base#previously_new_record?` to show if a record was new before the last save.
+
+    *Tom Ward*
+
 *   Allow generated `create_table` migrations to include or skip timestamps.
 
-     *Michael Duchemin*
+    *Michael Duchemin*
```
Comment (simi) [inline activerecord/CHANGELOG.md]: Looks addressed by https://github.com/rails/rails/commit/cc27e9988f9ce4e52acf491a5f67be70ab7d25d8#diff-1ed2907b3b8f148c2533558a77673ffaR17 already, I think you can :scissors: this from this PR.

**Comment by kaspth** [comment]
Thanks, @tomafro! 🙏

**Pull request merged**